File: backend/notification_service/producer/src/services/s3Service.js
const { S3Client, ListObjectsV2Command, DeleteObjectCommand, PutObjectCommand } = require('@aws-sdk/client-s3');
const config = require('../config/config');

class S3Service {
  constructor(bucketName = config.s3.bucketName) {
    this.client = new S3Client({
      credentials: {
        accessKeyId: config.s3.AccessKeyId,
        secretAccessKey: config.s3.SecretAccessKey,
      },
      region: config.s3.region,
    });
    this.bucketName = config.s3.bucketName;
  }

  // Listar objetos dentro del bucket
  async listObjects(prefix = '') {
    const params = {
      Bucket: this.bucketName,
      Prefix: prefix,
    };

    try {
      const command = new ListObjectsV2Command(params);
      const data = await this.client.send(command);
      return data.Contents ? data.Contents.map((obj) => obj.Key) : [];
    } catch (error) {
      console.error('Error listing objects from S3:', error.message);
      throw new Error('Error listing objects in S3');
    }
  }

  // Eliminar un objeto del bucket
  async deleteObject(objectKey) {
    const params = {
      Bucket: this.bucketName,
      Key: objectKey,
    };

    try {
      const command = new DeleteObjectCommand(params);
      await this.client.send(command);
      return true;
    } catch (error) {
      console.error('Error deleting object from S3:', error.message);
      throw new Error('Error deleting object from S3');
    }
  }

  // Subir archivo (equivalente a upload_object)
  async uploadObject(fileStream, objectKey) {
    const params = {
      Bucket: this.bucketName,
      Key: objectKey,
      Body: fileStream,
    };

    try {
      const command = new PutObjectCommand(params);
      await this.client.send(command);
      return `https://${this.bucketName}.s3.${config.s3.region}.amazonaws.com/${objectKey}`;
    } catch (error) {
      console.error('Error uploading file to S3:', error.message);
      throw new Error('Error uploading file to S3');
    }
  }
}

module.exports = new S3Service();

File: backend/app_service/app/services/s3_service.py
import boto3
from botocore.exceptions import NoCredentialsError, ClientError
import logging
from config import Config

class S3Service:
    def __init__(self, bucket_name=Config.AWS_BUCKET_NAME, region_name=Config.AWS_REGION):
        self.s3 = boto3.client('s3', region_name=region_name, aws_access_key_id=Config.AWS_ACCESS_KEY_ID, aws_secret_access_key=Config.AWS_SECRET_ACCESS_KEY)
        self.bucket_name = bucket_name

    def list_objects(self, prefix=''):
        try:
            response = self.s3.list_objects_v2(Bucket=self.bucket_name, Prefix=prefix)
            if 'Contents' in response:
                return [obj['Key'] for obj in response['Contents']]
            else:
                return []
        except NoCredentialsError:
            logging.error("Credentials not available.")
            return None
        except ClientError as e:
            logging.error(f"An error occurred: {e}")
            return None

    def delete_object(self, object_key):
        try:
            self.s3.delete_object(Bucket=self.bucket_name, Key=object_key)
            return True
        except NoCredentialsError:
            logging.error("Credentials not available.")
            return False
        except ClientError as e:
            logging.error(f"An error occurred: {e}")
            return False

    def upload_object(self, file_stream, object_key):
        try:
            # Usar upload_fileobj para archivos en memoria
            self.s3.upload_fileobj(file_stream, self.bucket_name, object_key)
            return f"https://{self.bucket_name}.s3.amazonaws.com/{object_key}"
        except NoCredentialsError:
            logging.error("Credentials not available.")
            return None
        except ClientError as e:
            logging.error(f"An error occurred: {e}")
            return None